Switch to native-routing in GKE guide

tgraf opened the pull request:
This changes the GKE installation instructions to default to direct-routing and provides the required documentation around it.

aanm commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -304,6 +304,36 @@ var _ = Describe("K8sDatapathConfig", func() {
+	Context("AutoDirectNodeRoutes", func() {
 		BeforeEach(func() {
 			SkipIfIntegration(helpers.CIIntegrationFlannel)
 			SkipIfIntegration(helpers.CIIntegrationGKE)
```
Does this still needs to be skipped?

tgraf replied:
Yes, a single L2 network is not given. In fact, auto-direct-node routes is not required at all on GKE as the default route will point to a router that can route all PodIPs as announced in Node.Spec.PodCIDR[s] so we can simply disable tunneling to achieve native routing across everything.

aanm commented on Documentation/gettingstarted/k8s-install-gke.rst:
```diff
@@ -72,6 +73,13 @@ correct one (see `GKE RBAC documentation`_ for more info).
+
+.. code:: bash
+
+    NATIVE_CIDR=$(gcloud container clusters describe $CLUSTER_NAME | grep clusterIpv4Cidr | awk '{print $2}')
```
small nit: `grep clusterIpv4Cidr -i`

joestringer commented on Documentation/gettingstarted/k8s-install-gke.rst:
```diff
@@ -88,7 +96,11 @@ below. This will ensure all pods are managed by Cilium.
+      --set global.k8s.requireIPv4PodCIDR=true \\
+      --set global.tunnel=disabled \\
+      --set global.endpointRoutes.enabled=true \\
+      --set global.native-routing-cidr=$NATIVE_CIDR
```
Do we also need to do something explicit about disabling cilium-health here? I think we'll allocate an address to it that's not routable.

tgraf replied:
The great news is that we don't have to. All IPs in the PodCIDR handed to us are fully-routable. The connectivity health checking is working as expected:
```
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.15+ (v1.15.11-gke.9) [linux/amd64]
Kubernetes APIs:        ["CustomResourceDefinition", "cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Endpoint", "core/v1::Namespace", "core/v1::Pods", "core/v1::Service", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Probe   [NodePort (HYBRID, 30000-32767, XDP: NONE), HostPort, ExternalIPs, HostReachableServices (TCP, UDP)]
Cilium:                 Ok      OK
NodeMonitor:            Disabled
Cilium health daemon:   Ok
IPAM:                   IPv4: 7/255 allocated from 10.68.2.0/24,
Controller Status:      36/36 healthy
Proxy Status:           OK, ip 10.68.2.184, 3 redirects active on ports 10000-20000
Hubble:                 Disabled
Cluster health:         3/3 reachable   (2020-04-21T20:52:18Z)
```

joestringer reviewed (changes requested):
One minor comment below.

tklauser commented on Documentation/concepts/datapath/gke.rst:
```diff
@@ -0,0 +1,65 @@
+Load-balancing
+   ClusterIP load-balancing will be performed using BPF for all version of GKE.
+   Starting with >= GKE v1.15 or when running a Linux kernel >= 4.19, all
+   NodePort/ExernalIP/HostPort will be performed using a BPF implementation as
```
Typo:
```suggestion
   NodePort/ExternalIP/HostPort will be performed using a BPF implementation as
```

tklauser commented on Documentation/concepts/datapath/gke.rst:
```diff
@@ -0,0 +1,65 @@
+* ``native-routing-cidr: x.x.x.x/y``: Set the CIDR in which native routing
+  is supported.
+
+See the getting started guide :ref:`k8s_install_gke` to install Cilium Google
```
Nit:

```suggestion
See the getting started guide :ref:`k8s_install_gke` to install Cilium on Google
```

tklauser commented on Documentation/concepts/datapath/native_routing.rst:
```diff
@@ -0,0 +1,69 @@
+Requirements on the network
+***************************
+
+* In order to run the native routing mode. The network connecting the hosts on
```
Nit: punctiuation

```suggestion
* In order to run the native routing mode, the network connecting the hosts on
```

tklauser commented on Documentation/concepts/datapath/native_routing.rst:
```diff
@@ -0,0 +1,69 @@
+private IP addresses. For any destination outside of the virtual network,
+traffic must typically be masqueraded. This is done by setting ``masquerade:
+true`` (default). In order to exclude the entire CIDR of the virtual network,
+the datapath must be told the CIDR within native routing is supported. This is
```
```suggestion
the datapath must be told the CIDR within which native routing is supported. This is
```

tklauser commented on Documentation/concepts/ipam/kubernetes.rst:
```diff
@@ -39,5 +45,21 @@ Annotation                          Description
+The following ConfigMap options exist to configure Kubernetes hostscope:
+
+ * ``ipam: kubernetes``: Enables Kubernetes IPAM mode. Enabling this option will
+   automatically enabled ``k8s-require-ipv4-pod-cidr`` if ``enable-ipv4`` is
```
```suggestion
   automatically enable ``k8s-require-ipv4-pod-cidr`` if ``enable-ipv4`` is
```

tklauser reviewed (commented):
Some small spelling/punctuation/typo remarks in the docs.

tgraf committed:
gke: Enable native-routing mode on GKE by default

The PodCIDR is always populated on GKE. Switch to native routing mode
while allocating IPs from Node.Spec.PodCIDR[s]

As we transition to native-routing mode:
 * We can enable endpoint routes to avoid traversing the cilium-host
   interface
 * We can disable the local node route as it is not required
 * We can disable local routing blacklisting as it is known that no
   local routes will conflict with the PodCIDR

Hide all of this behind a new option global.gke.enabled

Deriving the cluster CIDR as part of the guide also allows to disable
masquerading in the entire cluster.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Document GKE IPAM and datapath modes

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
test: Make remote-node identity disabled tests work with direct routing

When running native-routing mode, the source IP from host traffic is
unlikely the IP assigned to the cilium-host interface. In the
remote-node identity legacy mode, only the IP of the cilium-host
interface is considered host. All other IPs are considered world.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Remove clusterrolebinding step from GKE guide

This step does not seem required at this point. Remove it.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
test: Add coverage for GKE direct routing datapath modes

The GKE reference guide will switch over to using direct-routing by
default. Cover both modes in the GKE CI.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Correctly instruct to restart all unmanaged pods

Instead of instructing to only restart unmanaged pods in kube-system,
extend this to restart all unmanaged pods which are not running in host
networking mode.

Signed-off-by: Thomas Graf <thomas@cilium.io>
